src/tls.rs

```diff
@@ -1,6 +1,294 @@
+    fn as_raw_fd(&self) -> std::os::fd::RawFd {
+        let mut fd = -1;
+        let _ = unsafe { sys::esp_tls_get_conn_sockfd(self.reader.raw, &mut fd) };
+
+        fd
+    }
+}
+
+impl io::Io for EspTls {
+    type Error = EspIOError;
+}
+
+impl io::Read for EspTls {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, EspIOError> {
+        self.read(buf)
+    }
+}
+
+impl io::Write for EspTls {
+    fn write(&mut self, buf: &[u8]) -> Result<usize, EspIOError> {
+        self.write(buf)
+    }
+
+    fn flush(&mut self) -> Result<(), EspIOError> {
+        Ok(())
+    }
+}
+
+pub struct EspTlsRead {
+    raw: *mut sys::esp_tls,
+}
+
+impl EspTlsRead {
+    pub fn read(&mut self, buf: &mut [u8]) -> Result<usize, EspIOError> {
+        if buf.is_empty() {
+            return Ok(0);
+        }
+
+        // cannot call esp_tls_conn_read bc it's inline
+        let esp_tls = unsafe { std::ptr::read_unaligned(self.raw) };
+        let read_func = esp_tls.read.unwrap();
+        let ret = unsafe { read_func(self.raw, buf.as_mut_ptr() as *mut i8, buf.len()) };
+        // ESP docs treat 0 as error, but in Rust it's common to return 0 from `Read::read` to indicate eof
+        if ret >= 0 {
+            Ok(ret as usize)
+        } else {
+            Err(EspIOError(EspError::from(ret as i32).unwrap()))
+        }
+    }
+}
+
+impl io::Io for EspTlsRead {
+    type Error = EspIOError;
+}
+
+impl io::Read for EspTlsRead {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, EspIOError> {
+        self.read(buf)
+    }
+}
+
+pub struct EspTlsWrite {
+    raw: *mut sys::esp_tls,
+}
+
+impl EspTlsWrite {
+    pub fn write(&mut self, buf: &[u8]) -> Result<usize, EspIOError> {
+        if buf.is_empty() {
+            return Ok(0);
+        }
+
+        // cannot call esp_tls_conn_write bc it's inline
+        let esp_tls = unsafe { std::ptr::read_unaligned(self.raw) };
+        let write_func = esp_tls.write.unwrap();
+        let ret = unsafe { write_func(self.raw, buf.as_ptr() as *const i8, buf.len()) };
+        if ret >= 0 {
+            Ok(ret as usize)
+        } else {
+            Err(EspIOError(EspError::from(ret as i32).unwrap()))
+        }
+    }
+}
+
+impl io::Io for EspTlsWrite {
+    type Error = EspIOError;
+}
+
+impl io::Write for EspTlsWrite {
+    fn write(&mut self, buf: &[u8]) -> Result<usize, EspIOError> {
+        self.write(buf)
+    }
+
+    fn flush(&mut self) -> Result<(), EspIOError> {
+        Ok(())
+    }
+}
+
+impl Drop for EspTls {
+    fn drop(&mut self) {
+        unsafe {
+            sys::esp_tls_conn_destroy(self.reader.raw);
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct Config<'a> {
+    /// up to 9 ALPNs allowed, with avg 10 bytes for each name
+    pub alpn_protos: Option<&'a [&'a str]>,
+    pub ca_cert: Option<X509<'a>>,
+    pub client_cert: Option<X509<'a>>,
+    pub client_key: Option<X509<'a>>,
+    pub client_key_password: Option<&'a str>,
+    pub non_block: bool,
```

Comment:
If this is `true`, should we call `conn_new_async` instead?

Reply:
Not necessarily, but **VERY IMPORTANT**: I was [looking at the code]() and I think I just figured out a way for us to pass an already opened socket!

I think we still should keep the existing `new`, but we should introduce (maybe post-merge) a new STD-only method along the lines of
```rust
#[cfg(feature = "std")]
pub fn wrap(socket: std::net::TcpSocket, cfg: &Config) -> Result<EspError, Self> {
    let tls = esp!(unsafe { esp_tls_init(); })?;
    esp!(unsafe { esp_tls_set_conn_state(tls, esp_tls_conn_state_t_ESP_TLS_CONNECTING) })?;
    esp!(unsafe { esp_tls_set_conn_sockfd(tls, socket.fd)? })?;
    esp!(unsafe { esp_tls_conn_new_sync(b"\0", 0, 0, cfg, tls) })?;

    Ok(...)
}
```

The above ^^^ will not try to open a new socket and will re-use ours, as the state of the connection is `esp_tls_conn_state_t_ESP_TLS_CONNECTING`.

The above approach is important, because it might be a way to implement `async` as well. Again, to do `async` in a way which is independent of the concrete async Reactor, we needed a way to pass-in an already opened socket handle, and the above might be just this.

Reply:
To not make the above STD-only, we can even pass to the method a raw file descriptor, where the assumption is that the file descriptor is a socket descriptor, and also it is already opened.